### PR TITLE
Support for Argo backend

### PR DESCRIPTION
**Pull request by Nilabhra**
### Additions:
- Support for executing pipelines as Argo Workflows on Kubernetes.
- Pipelines would be represented as DAGs, making pipeline representations generic.
- `.dockerignore` file to filter out unneeded files from copied over to the container. Currently, everything in the project root get's copied over, including inputs and outputs.

### Changes:
- Modification of the `dockerize` module to support tagging + pushing a docker image to Docker Hub. For local deployments, later, support for a local docker registry needs to be added.
- Modification of the `example.py` file to not accept the transaction model file name directly. The assumption now is that the user would provide the path to the directory where the pre-trained model binary is present. This had to be done as Kubernetes doesn't allow mounting of files as volumes. This modification seemed reasonable to me since most saved models use a unique directory name with a generic name for the weights binary (such as `model.bin` or `weights.pt`).
- I was unable to get the shell scripts running as they were.  I decided to just use `python`, which seems to work for me and @nikhil.
- The `dockerize` module avoids setting the entry point to `conda run -n pircli`. While this works perfectly for docker-compose, it fails to work for Argo. I decided to just set the PATH env var to the location of the python binary for the `pircli` environment. That worked both for docker and for Argo.

### Closes:
- #18 

### Comments:
I have tested out both the Argo backend the existing Docker backend and ensure they both work with the recent changes.

### Commit history:
* add: subparser for argoize

* add: Nikhil's changes.

* feat: generating workflow name using output file name.

* feat: function to create Argo template from node.

* chore: added docstring.

* chore: added comments.

* feat: setting the env vars before generation of the argo yaml.

* feat: function to generate NFS volume specs for K8s.

* feat: added support for mounting NFS volumes.

* fix: using absolute path for the NFS mount paths.

* fix: typo.

* feat: function for generating Argo templates for PIRlib Graph objects.

* chore: added comments.

* feat: enabled creation of DAG tasks.

* chore: added Argo specific refactoring.

* chore: refactored argo task names.

* fix: files cannot be mounted as volumes in Argo/K8s.

* chore: removed redundant module.

* chore: debugging code.

* chore: optimized docker image generation.

* add: dockerignore file to prevent copying of unnecessary files.

* fix: missing volume mount.

* chore: reverted to original formatting.

* chore: removed spurious code.

* chore: increased black linelength to 100.

### CODE REVIEW

**Comment by zhanyuanucb** [comment]
Can we include the requirements to run the new Argo example?

Also, we can include the generated YAML files from the new example, like docker and local backends

If this makes the `example` dir looks messy, we can organize different example in different folders.

File: pirlib/backends/argo_batch.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    dependencies = []
+
+    # If the node has a valid graph input source,
+    # mount the respective host system's file to the input volume
+    for inp in node.inputs:
+        if inp.source.graph_input_id:
+            # Construct the env var name for the env var that contains the
+            # input directory/file path.
+            inp_name = inp.source.graph_input_id
+            inp_env_var = f"INPUT_{inp_name}"
```
Comment (zhanyuanucb) [inline pirlib/backends/argo_batch.py:92]: Add error message for not set env vars.
Something like `f”${{INPUT_{inp_id}?INPUT_{inp_id} not found}}”` should be fine

File: pirlib/backends/argo_batch.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+    """
+    name = graph.id
+    image = graph.nodes[0].entrypoints["main"].image
+    command = ["python", "-m", __name__, "graph", graph_outputs_encoded]
+    volumes = [create_nfs_volume_spec("node_outputs", "OUTPUT")]
+    volume_mounts = [{"name": argo_name("node_outputs"), "mountPath": "/mnt/node_outputs"}]
+
+    for inp in graph.inputs:
+        inp_name = inp.id
+        inp_env_var = f"INPUT_{inp_name}"
```
Comment (zhanyuanucb) [inline pirlib/backends/argo_batch.py:148]: Same as above, add some error message for not able to find env vars

File: pirlib/backends/argo_batch.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    volume_mounts = [{"name": argo_name("node_outputs"), "mountPath": "/mnt/node_outputs"}]
+    dependencies = []
+
+    # If the node has a valid graph input source,
+    # mount the respective host system's file to the input volume
+    for inp in node.inputs:
+        if inp.source.graph_input_id:
+            # Construct the env var name for the env var that contains the
+            # input directory/file path.
+            inp_name = inp.source.graph_input_id
```
Comment (zhanyuanucb) [inline pirlib/backends/argo_batch.py:91]: may be `inp_id = inp.source.graph_input_id` will be better?

File: pirlib/cli/dockerize.py [hunk 4]
```diff
@@ -90,13 +113,12 @@ def _generate_dockerfile(context_path: pathlib.Path) -> str:
                 "RUN wget https://repo.anaconda.com/miniconda/Miniconda3"
                 "-latest-Linux-$(uname -m).sh -O /tmp/Miniconda3.sh"
             ),
-            f"RUN bash /tmp/Miniconda3.sh -b -p {miniconda3}",
+            f"RUN bash /tmp/Miniconda3.sh -b -p {miniconda3}",  # && rm /tmp/Miniconda3.sh",
```
Comment (zhanyuanucb) [inline pirlib/cli/dockerize.py:116]: If we do need this `rm`, we can uncomment it, otherwise we can remove it

File: pirlib/cli/dockerize.py [hunk 5]
```diff
@@ -90,13 +113,12 @@ def _generate_dockerfile(context_path: pathlib.Path) -> str:
...
             "RUN echo $CONDA_ENV_B64 | base64 -d > /tmp/environment.yml",
             f"RUN {conda} env create -n pircli -f /tmp/environment.yml",
-            f'ENTRYPOINT ["{conda}", "run", "-n", "pircli"]',
-            f"ENV PYTHONPATH={pythonpath}",
             f"COPY . {workdir}",
             f"WORKDIR {workdir}",
+            f'ENV PATH="{miniconda3}/envs/pircli/bin"',
```
Comment (zhanyuanucb) [inline pirlib/cli/dockerize.py:121]: Why did we change from `PYTHONPATH` to `PATH` here? `PATH` is for linux system to search for executables?  I think here we want to use `PYTHONPATH` so that whatever is in users’ local `PYTHONPATH` (which is also in the build context, see `_infer_pythonpath`) can be prepended to `sys.path`

File: example/run_argo.sh [hunk 6]
```diff
@@ -0,0 +1,25 @@
+EXAMPLEDIR=$(dirname $0)
+ROOTDIR=$EXAMPLEDIR/..
+
+### Module 1: Docker_Packaging
+python $ROOTDIR/bin/pircli dockerize \
+    --auto $ROOTDIR \
```
Comment (zhanyuanucb) [inline example/run_argo.sh:6]: TBH, `—auto $ROOTDIR` is a bit confusing.
May be something like this will be better?
```
$ROOTDIR
--auto
```

File: pirlib/backends/argo_batch.py (hunk 1 above)
Comment (Nilabhra) [reply pirlib/backends/argo_batch.py:92]: Sure, will add this.

File: pirlib/cli/dockerize.py (hunk 4 above)
Comment (Nilabhra) [reply pirlib/cli/dockerize.py:116]: We don't need it at the moment, but going forward, it would be nice to have it included as it would delete the shell file and thereby reduce the image size by almost 70MB. At the moment, it's commented out as it makes development faster. Let me know what you think.

File: pirlib/cli/dockerize.py (hunk 5 above)
Comment (Nilabhra) [reply pirlib/cli/dockerize.py:121]: @zhanyuanucb 
The existing implementation with `conda run` was not working and the system kept using the Python binary in `/usr/bin`. On looking for answers online, I came across the fact that people have reported that it's a hit and miss as `conda run` is still experimental.

In this PR I reset `PATH` so that the system finds the Python binary in the `pircli` env. This of course is not a good solution as the other system binaries cannot be found this way. In the next PR, I prepended the path to the Python binary to `PATH` for more stability.

With regards to `PYTHONPATH` I can add it back. I guess it was added so that modules under development (or the modules inside `examples`) could be imported. But since `pirlib` was installed with the `-e` flag of pip, it wasn't a requirement.

File: example/run_argo.sh (hunk 6 above)
Comment (Nilabhra) [reply example/run_argo.sh:6]: Sure!

**Comment by Nilabhra** [comment]
@zhanyuanucb 
>Can we include the requirements to run the new Argo example?
>Also, we can include the generated YAML files from the new example, like docker and local backends
>If this makes the example dir looks messy, we can organize different example in different folders.

The issue with this is that it would expose specific information about my development setup such as:
- My DockerHub name.
- The name of the DockerHub repo.
- The file locations location of the inputs/outputs e.g. `/home/nilabhra/pirlib/example/inputs/train_dataset`.

None, of this is sensitive, so i have no issue exposing this if you are fine with it.

File: pirlib/backends/argo_batch.py (hunk 2 above)
Comment (Nilabhra) [reply pirlib/backends/argo_batch.py:148]: Handling this and the previous one under `create_nfs_volume_spec`

File: pirlib/backends/argo_batch.py (hunk 1 above)
Comment (zhanyuanucb) [reply pirlib/backends/argo_batch.py:92]: right

File: pirlib/cli/dockerize.py (hunk 5 above)
Comment (zhanyuanucb) [reply pirlib/cli/dockerize.py:121]: Got it. I didn't know that `conda run` has this bug. Could you point me to the link to the webpage describing this issue?

Could you prepend the path to `PATH` in this PR so that we don't need to wait for the next PR?

I think we still need the prepend the `PYTHONPATH` through `_infer_pythonpath` since the build context may not necessarily be the root of `pirlib` and it could be any working directory of users. I think `_infer_pythonpath` assure that users' packages that are not installed via `pip` can be found by Python in the container.

**Comment by zhanyuanucb** [comment]
@Nilabhra 
For those personal information, you can use something like `xxx` or `@@@` to mask it out

> @zhanyuanucb
> 
> > Can we include the requirements to run the new Argo example?
> 
> Sure, will add that.
> 
> > Also, we can include the generated YAML files from the new example, like docker and local backends
> > If this makes the example dir looks messy, we can organize different example in different folders.
> 
> The issue with this is that it would expose specific information about my development setup such as:
> 
> * My DockerHub name.
> * The name of the DockerHub repo.
> * The file locations location of the inputs/outputs e.g. `/home/nilabhra/pirlib/example/inputs/train_dataset`.
> 
> None of this is sensitive, so i have no issue exposing this if you are fine with it.

File: pirlib/cli/dockerize.py (hunk 4 above)
Comment (Nilabhra) [reply pirlib/cli/dockerize.py:116]: Sure I can try that.

File: pirlib/cli/dockerize.py (hunk 5 above)
Comment (Nilabhra) [reply pirlib/cli/dockerize.py:121]: > Got it. I didn't know that conda run has this bug. Could you point me to the webpage describing this issue?

- This is the blog post that mentions that `conda run` is "experimental" https://pythonspeed.com/articles/activate-conda-dockerfile/
- This is a long GitHub issue thread where people mention that it doesn't always work for them: https://github.com/ContinuumIO/docker-images/issues/89

> Could you prepend the path to PATH in this PR so that we don't need to wait for the next PR? 

Sure. Once this PR is accepted, I'll merge and rebase all the changes to the next PR (#21) from the master branch.

> I think we still need the prepend the PYTHONPATH through _infer_pythonpath since the build context may not necessarily be the root of pirlib and it could be any working directory of users. I think _infer_pythonpath assures that users' packages that are not installed via pip can be found by Python in the container.

Good point! I didn't think of the scenario where there might be code residing outside the `pirlib` root that would need to be imported. I'll add back `PYTHONPATH`.

**Comment by Nilabhra** [comment]
@zhanyuanucb 
> @Nilabhra
For those personal information, you can use something like xxx or @@@ to mask it out

Sure, just be aware that the YAML files would not work out of the box unlike the YAML file for docker-compose.

File: pirlib/cli/dockerize.py (hunk 4 above)
Comment (Nilabhra) [reply pirlib/cli/dockerize.py:116]: @zhanyuanucb The miniconda image works, hence there is no need to install and then remove the Miniconda shell file anymore. I was expecting `conda run` would also work, but it wasn't the case. Prepending the env location to `$PATH` seems to work though.

**Comment by Nilabhra** [comment]
@zhanyuanucb Up for another round of review.

**Comment by zhanyuanucb** [comment]
@Nilabhra Could you put all the examples together in one .md file so that reader can see the toy example running on different backends? Maybe later the same .md file can be extended by the `wiki_parser` example, which is shown as an advance example.

The rest looks good to me.

**Comment by Nilabhra** [comment]
@zhanyuanucb 
> Could you put all the examples together in one .md file so that reader can see the toy example running on different backends? Maybe later the same .md file can be extended by the wiki_parser example, which is shown as an advance example.

I can copy-paste some of the content of the [`README.rst`](https://github.com/Nilabhra/pirlib/blob/master/docs/README.rst) in the docs directory. Would that be fine? Or should I remove the  **Examples** section from existing `README.rst` too?
